In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("data/raw/churn.csv")

In [4]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [5]:
df.shape

(7043, 21)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## Data Info
Row: 7043  
Col: 21  
Numeric:
- SeniorCitizen (int)
- tenure (int)
- MonthlyCharges (float)  

Object:
- others?

### Identify categorical columns

In [28]:
obj_col = df.select_dtypes(include=['object']).columns.tolist()
num_col = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
print(obj_col)
print(num_col)

['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']
['SeniorCitizen', 'tenure', 'MonthlyCharges']


In [67]:
# Get each Object Column's unique data and number of unique of data
categorical_col = []
numerical_col = []
for col in obj_col:
    print(f" - {col} \t: {df[col].nunique()} \t, {df[col].unique()}")
    if df[col].nunique() < 10:
        categorical_col.append(col)
    elif df[col].nunique() < df.shape[0]:
        numerical_col.append(col)
print(f"\n\nCategorical Columns: {categorical_col}")
print(f"\n\nNumerical Columns: {numerical_col}")


 - customerID 	: 7043 	, ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
 - gender 	: 2 	, ['Female' 'Male']
 - Partner 	: 2 	, ['Yes' 'No']
 - Dependents 	: 2 	, ['No' 'Yes']
 - PhoneService 	: 2 	, ['No' 'Yes']
 - MultipleLines 	: 3 	, ['No phone service' 'No' 'Yes']
 - InternetService 	: 3 	, ['DSL' 'Fiber optic' 'No']
 - OnlineSecurity 	: 3 	, ['No' 'Yes' 'No internet service']
 - OnlineBackup 	: 3 	, ['Yes' 'No' 'No internet service']
 - DeviceProtection 	: 3 	, ['No' 'Yes' 'No internet service']
 - TechSupport 	: 3 	, ['No' 'Yes' 'No internet service']
 - StreamingTV 	: 3 	, ['No' 'Yes' 'No internet service']
 - StreamingMovies 	: 3 	, ['No' 'Yes' 'No internet service']
 - Contract 	: 3 	, ['Month-to-month' 'One year' 'Two year']
 - PaperlessBilling 	: 2 	, ['Yes' 'No']
 - PaymentMethod 	: 4 	, ['Electronic check' 'Mailed check' 'Bank transfer (automatic)'
 'Credit card (automatic)']
 - TotalCharges 	: 6531 	, ['29.85' '1889.5' '108.15' ... '3

In [68]:
for col in num_col:
    print(f" - {col} \t: {df[col].nunique()} \t, {df[col].unique()}")
    if df[col].nunique() < 10:
        categorical_col.append(col)
    elif df[col].nunique() < df.shape[0]:
        numerical_col.append(col)
print(f"\n\nCategorical Columns: {categorical_col}")
print(f"\n\nNumerical Columns: {numerical_col}")

 - SeniorCitizen 	: 2 	, [0 1]
 - tenure 	: 73 	, [ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]
 - MonthlyCharges 	: 1585 	, [29.85 56.95 53.85 ... 63.1  44.2  78.7 ]


Categorical Columns: ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn', 'SeniorCitizen']


Numerical Columns: ['TotalCharges', 'tenure', 'MonthlyCharges']


In [69]:
for col in categorical_col:
    print(f"\n - {col} \t:")
    for category in df[col].unique():
        print(f"\t - {category}: {sum(df[col]==category)}, percentage: {round(sum(df[col]==category)/df.shape[0],2)}")


 - gender 	:
	 - Female: 3488, percentage: 0.5
	 - Male: 3555, percentage: 0.5

 - Partner 	:
	 - Yes: 3402, percentage: 0.48
	 - No: 3641, percentage: 0.52

 - Dependents 	:
	 - No: 4933, percentage: 0.7
	 - Yes: 2110, percentage: 0.3

 - PhoneService 	:
	 - No: 682, percentage: 0.1
	 - Yes: 6361, percentage: 0.9

 - MultipleLines 	:
	 - No phone service: 682, percentage: 0.1
	 - No: 3390, percentage: 0.48
	 - Yes: 2971, percentage: 0.42

 - InternetService 	:
	 - DSL: 2421, percentage: 0.34
	 - Fiber optic: 3096, percentage: 0.44
	 - No: 1526, percentage: 0.22

 - OnlineSecurity 	:
	 - No: 3498, percentage: 0.5
	 - Yes: 2019, percentage: 0.29
	 - No internet service: 1526, percentage: 0.22

 - OnlineBackup 	:
	 - Yes: 2429, percentage: 0.34
	 - No: 3088, percentage: 0.44
	 - No internet service: 1526, percentage: 0.22

 - DeviceProtection 	:
	 - No: 3095, percentage: 0.44
	 - Yes: 2422, percentage: 0.34
	 - No internet service: 1526, percentage: 0.22

 - TechSupport 	:
	 - No: 3473,

In [71]:
print(df.shape[0])
for i in range(df.shape[0]):
    try:
        if float(df["TotalCharges"][i]):
            continue
        # print(f" - TotalCharges \t: ",float(df["TotalCharges"][i]))
    except:
        print(f"In {i}, value: ({df['TotalCharges'][i]}) can't convert to float")

7043
In 488, value: ( ) can't convert to float
In 753, value: ( ) can't convert to float
In 936, value: ( ) can't convert to float
In 1082, value: ( ) can't convert to float
In 1340, value: ( ) can't convert to float
In 3331, value: ( ) can't convert to float
In 3826, value: ( ) can't convert to float
In 4380, value: ( ) can't convert to float
In 5218, value: ( ) can't convert to float
In 6670, value: ( ) can't convert to float
In 6754, value: ( ) can't convert to float


## Data Info
Data Source: Telco Customer Churn from Kaggle/IBM Sample data
Download time: 2026-01-11
Label Leakage: since it is from Kaggle for practice, there might be some leakage
Row: 7043  (Each row represents a single customer snapshot)
Col: 21  

## Columns list

Primary Key:  
- customerID  : unique 7043 value should be the primary key and unique for each row. There's no time related columns in the dataset

Numeric:
- SeniorCitizen (int) (Note: the original data info is int64, and only include 2 values ie, 0 and 1. Should concider as categorical)
- tenure (int) (Note: original data info is int64, and include 73 unique values. Should concider as categorical)
- MonthlyCharges (float)
- TotalCharges (object) (Note: original object as String type, but need to convert to float. There's some missing data as value ' ' (blank space))

Categorical:  
(Note 1: in multiple columns, they include "No internet servie" might have high relation with InternetService. Should investigate the correlation)  
(Note 2: in MultipleLines, "No phone service" is align with PhoneService's "No" which is not missing value and have strong correlation between those columns)
 - gender 	:
	 - Female: 3488, percentage: 0.5
	 - Male: 3555, percentage: 0.5

 - Partner 	:
	 - Yes: 3402, percentage: 0.48
	 - No: 3641, percentage: 0.52

 - Dependents 	:
	 - No: 4933, percentage: 0.7
	 - Yes: 2110, percentage: 0.3

 - PhoneService 	:
	 - No: 682, percentage: 0.1
	 - Yes: 6361, percentage: 0.9

 - MultipleLines 	: 
	 - No phone service: 682, percentage: 0.1
	 - No: 3390, percentage: 0.48
	 - Yes: 2971, percentage: 0.42

 - InternetService 	:
	 - DSL: 2421, percentage: 0.34
	 - Fiber optic: 3096, percentage: 0.44
	 - No: 1526, percentage: 0.22

 - OnlineSecurity 	:
	 - No: 3498, percentage: 0.5
	 - Yes: 2019, percentage: 0.29
	 - No internet service: 1526, percentage: 0.22

 - OnlineBackup 	:
	 - Yes: 2429, percentage: 0.34
	 - No: 3088, percentage: 0.44
	 - No internet service: 1526, percentage: 0.22

 - DeviceProtection 	:
	 - No: 3095, percentage: 0.44
	 - Yes: 2422, percentage: 0.34
	 - No internet service: 1526, percentage: 0.22

 - TechSupport 	:
	 - No: 3473, percentage: 0.49
	 - Yes: 2044, percentage: 0.29
	 - No internet service: 1526, percentage: 0.22

 - StreamingTV 	:
	 - No: 2810, percentage: 0.4
	 - Yes: 2707, percentage: 0.38
	 - No internet service: 1526, percentage: 0.22

 - StreamingMovies 	:
	 - No: 2785, percentage: 0.4
	 - Yes: 2732, percentage: 0.39
	 - No internet service: 1526, percentage: 0.22

 - Contract 	:
	 - Month-to-month: 3875, percentage: 0.55
	 - One year: 1473, percentage: 0.21
	 - Two year: 1695, percentage: 0.24

 - PaperlessBilling 	:
	 - Yes: 4171, percentage: 0.59
	 - No: 2872, percentage: 0.41

 - PaymentMethod 	:
	 - Electronic check: 2365, percentage: 0.34
	 - Mailed check: 1612, percentage: 0.23
	 - Bank transfer (automatic): 1544, percentage: 0.22
	 - Credit card (automatic): 1522, percentage: 0.22

 - Churn 	:
	 - No: 5174, percentage: 0.73
	 - Yes: 1869, percentage: 0.27

 - SeniorCitizen 	:
	 - 0: 5901, percentage: 0.84
	 - 1: 1142, percentage: 0.16
